### ISSUE TITLE
[Design Request] Adding Affordance to Pagination

### ISSUE THREAD

**Issue by stuartjmoore** [MiSTer-devel/Main_MiSTer#926]
As a new user, I wanted to share some ideas I had after booting up the system for the first time. I’ve grown to really appreciate the UX/IA behind the design; these thoughts are only to enhance what’s already there.

Navigating the menu took me some time to figure out, primarily due to issues with affordance—specifically, understanding what each button does before pressing it. Though, different controller layouts and the limited hardware buttons certainly contribute to the confusion.

One of the most perplexing actions was horizontally paging between menu screens. While I appreciate this feature, tapping left and suddenly seeing all the options change is quite jarring. The only indication of this action is the arrows on the “exit” row at the very bottom, which don’t clearly convey their function.

### Concepts

A simple and familiar solution, reminiscent of the Xbox’s blade UI, would be to display the adjacent menu titles on the left and right sides of the OSD. This would help users see their current position and potential navigation options. The existing left and right arrows (icons in the font) could indicate which buttons to press to navigate.

![blades](https://github.com/user-attachments/assets/aaf6d567-d2d3-4ecc-8330-faa007f7d831)

An alternative and simpler design would include pagination indicator dots below the OSD, similar to the ones in iOS. The bounding box would be custom drawn, but the filled and outlined boxes are already included in the font. While it may not be as explicit as the blade UI, it would provide users with a better sense of their location within the menu.

![indicators](https://github.com/user-attachments/assets/2a374bd0-ac34-4573-8b52-d0dfce9e3363)

### Implementation

While this would add some complexity to the OSD to have to define which menus are adjacent to each one, a quick implementation could repurpose and move the “exit” row arrows:

![arrows](https://github.com/user-attachments/assets/71269cf8-e23a-4e73-8ed6-d38115491d56)

**Comment by birdybro** [comment]
I really like this idea. New users are often confused about there being more to the menu than they are seeing and this would give them an intuitive signal on what to do or what to try. :)

**Comment by SeongGino** [comment]
Hrm... would it be possible to use the bottommost line (or otherwise extend the OSD by one line) for page indication?

Assuming that physical size of the layout is the limitation (I'm not personally aware of what the technical "limits" of current OSD are), I'm sure even this would be a good indicator.
![menu indicators bottom](https://github.com/user-attachments/assets/c6f25e22-5588-447a-8290-590dbb675d9f)

Not sure about colors, might be limited by how long the console name is, but the idea is to either change the highlight without moving the items, or shift the items so that the current selection is always in the center.

**Comment by stuartjmoore** [comment]
@SeongGino similar to today, but sticky. I like it. I’ll spec out the pixels and add it to the list.

It’s possible that we could do both based on the amount of space available. Even if you’re using NTSC, you should have ~480 pixels to play with, but don’t quote me on that.

**Comment by SeongGino** [comment]
> @SeongGino similar to today, but sticky. I like it. I’ll spec out the pixels and add it to the list.
> 
> It’s possible that we could do both based on the amount of space available. Even if you’re using NTSC, you should have ~480 pixels to play with, but don’t quote me on that.

Hey, thanks! :)

To be fair, I don't think literal resolution is the limiting factor here--IIRC I'd read something from (I think) Sorg's recounts that there's not a lot of resources to go around with the current OSD as it is, which is why it's so functionally/aesthetically simple. Not sure if it's memory or (current) layout limits or what... but don't quote me on that. c;

**Comment by sorgelig** [comment]
OSD can output only inside existing square. Any outside element will require additional precious BRAM resource, so it's no-go option.
Adding some footer/header inside existing square is possible but it's not super helpful because MiSTer uses only 2 level menu and current sub-menu name is displayed on the left side. It's also optimized for simple navigation by 4 directional key and 2 buttons. Adding additional navigation will require more complicated key combos which is not really useful.

**Comment by stuartjmoore** [comment]
> Adding some footer/header inside existing square is possible but it's not super helpful because MiSTer uses only 2 level menu and current sub-menu name is displayed on the left side.

This isn’t for sub-menus, it’s for the lateral menus at the same level, but aren’t indicated anywhere.

> It's also optimized for simple navigation by 4 directional key and 2 buttons. Adding additional navigation will require more complicated key combos which is not really useful.

This doesn’t change any navigation or input, it only displays what existing keys will do when pressed.

**Comment by SeongGino** [comment]
Pretty much agreed with Stuart; as it is without any indicators, it's hard to even know there *is* a sub-menu in the first place without being explicitly told about it beforehand. And even then, just being made aware of what the other screen is adjacent to the one we're on would be helpful.

I guess my proposal as it was drawn would be slightly redundant showing the selected page name at the bottom when the sidebar does that as well, but it would be nice to have it show what page being switched to correlates to what direction is being pressed on the input device/pad.

It's understandable that kind of info might not be prudent for those already experienced with MiSTer--but for the however-many-other-thousands of people jumping in via cheaper board options as of late, it would help to smooth over the UX a little with some basic labeling.

**Comment by sorgelig** [comment]
> it's hard to even know there _is_ a sub-menu in the first place without being explicitly told about it beforehand.

Every sub-menu is marked by triangle on the right. Pretty much distinguishable.

You are welcome to propose changes through pull request. I mean not drawing the pictures but modifying existing code to the way you think will be better.

**Comment by stuartjmoore** [comment]
> Every sub-menu is marked by triangle on the right. Pretty much distinguishable.

Yes, but this isn’t for the sub-menu, it’s for the menus to the left and right of the one you’re currently in. The ones that aren’t marked anywhere.

> You are welcome to propose changes through pull request. I mean not drawing the pictures but modifying existing code to the way you think will be better.

This _is_ the proposal and RFC. We work out the design and ideas for someone looking to pick up work. There’s no reason to put in the coding effort if it’ll be rejected or need heavy modification.

**Comment by sorgelig** [comment]
on the last line of menu you can see < and > marking availability either page on the left or right.
